### PR TITLE
fix(write-api): Partial writes now return a 400 response code

### PR DESCRIPTION
**Pull request by savage-engineer**
Hi folks, this will help with https://github.com/influxdata/influxdb_iox/issues/11897.

The response code for partial writes is being changed to 400 imminently.

I've updated the `PostWrite` documentation for dedicated and serverless, as well as the troubleshooting points tables.

Thanks in advance 🙏 

---

* fix(write-api): Serverless write requests outside retention return 400 (e448b6a22)

      Partial writes are the only mode of operation on serverless clusters.
      This means that any requests where all points are outside the retention
      policy of a bucket receive a 400 bad request.

* fix(write-api): Partial writes now return a 400 response code in serverless (8305b7cde)

      This updates some of the other wording around the troubleshooting table
      to reflect the product's current state more closely.

* fix(write-api): Partial writes now return a 400 response code in dedicated (5d74f9792)

      This updates some of the other wording around the troubleshooting table
      to reflect the product's current state more closely.

### CODE REVIEW

File: api-docs/cloud-dedicated/v2/ref.yml [hunk 1]
```diff
@@ -656,33 +656,29 @@ paths:
...
-                    message: 'failed to parse line protocol: errors encountered on line(s): error message for first rejected point</n> error message for second rejected point</n> error message for Nth rejected point (up to 100 rejected points)'
-              schema:
-                $ref: '#/components/schemas/LineProtocolError'
         '204':
           description: Success ("No Content"). All data in the batch is written and queryable.
         '400':
-          description: Bad Request. All data in body was rejected and not written.
+          description: |
+            Data from the batch was rejected and not written. The response body indicates if a partial write occurred or all data was rejected.
+            If a partial write occurred then some points from the batch are written and queryable.
```
Comment (jstirnaman) [inline api-docs/cloud-dedicated/v2/ref.yml:664]:
```suggestion
            If a partial write occurred, then some points from the batch are written and queryable. 
```

File: api-docs/cloud-dedicated/v2/ref.yml [hunk 2]
```diff
@@ -656,33 +656,29 @@ paths:
...
-                $ref: '#/components/schemas/LineProtocolError'
         '204':
           description: Success ("No Content"). All data in the batch is written and queryable.
         '400':
-          description: Bad Request. All data in body was rejected and not written.
+          description: |
+            Data from the batch was rejected and not written. The response body indicates if a partial write occurred or all data was rejected.
+            If a partial write occurred then some points from the batch are written and queryable. 
+
+            The response body contains details about the [rejected points](/influxdb/cloud-serverless/write-data/troubleshoot/#troubleshoot-rejected-points), up to 100 points.
```
Comment (jstirnaman) [inline api-docs/cloud-dedicated/v2/ref.yml:666]:
```suggestion
            The response body contains details about the [rejected points](/influxdb/cloud-dedicated/write-data/troubleshoot/#troubleshoot-rejected-points), up to 100 points.
```

File: api-docs/cloud-serverless/v2/ref.yml [hunk 3]
```diff
@@ -7639,27 +7639,14 @@ paths:
...
-                    message: 'failed to parse line protocol: errors encountered on line(s): error message for first rejected point</n> error message for second rejected point</n> error message for Nth rejected point (up to 100 rejected points)'
-              schema:
-                $ref: '#/components/schemas/LineProtocolError'
         '204':
           description: Success ("No Content"). All data in the batch is written and queryable.
         '400':
           description: |
-            All data in the batch was rejected and not written.
+            Data from the batch was rejected and not written. The response body indicates if a partial write occurred or all data was rejected.
+            If a partial write occurred then some points from the batch are written and queryable.
```
Comment (jstirnaman) [inline api-docs/cloud-serverless/v2/ref.yml:7647]:
```suggestion
            If a partial write occurred, then some points from the batch are written and queryable. 
```

File: content/influxdb/cloud-dedicated/write-data/troubleshoot.md [hunk 4]
```diff
@@ -33,8 +33,7 @@ Learn how to avoid unexpected results and recover from errors when writing to {{
   3. Ingests or rejects data in the batch and returns one of the following HTTP status codes:
 
      - `204 No Content`: all data in the batch is ingested
-     - `201 Created` (_If the cluster is configured to allow **partial writes**_): some points in the batch are ingested and queryable, and some points are rejected
-     - `400 Bad Request`: all data is rejected
+     - `400 Bad Request`: some (_When **partial writes** are configured for the cluster_) or all of the data has been rejected. data which has not been rejected is ingested and queryable
```
Comment (jstirnaman) [inline content/influxdb/cloud-dedicated/write-data/troubleshoot.md:36]:
```suggestion
     - `400 Bad Request`: Some (_when **partial writes** are configured for the cluster_) or all of the data has been rejected. Data that has not been rejected is ingested and queryable.
```

File: content/influxdb/cloud-dedicated/write-data/troubleshoot.md [hunk 5]
```diff
@@ -33,8 +33,7 @@ Learn how to avoid unexpected results and recover from errors when writing to {{
   3. Ingests or rejects data in the batch and returns one of the following HTTP status codes:
 
      - `204 No Content`: all data in the batch is ingested
```
Comment (jstirnaman) [inline content/influxdb/cloud-dedicated/write-data/troubleshoot.md:35]:
```suggestion
     - `204 No Content`: All data in the batch is ingested.
```

File: content/influxdb/cloud-dedicated/write-data/troubleshoot.md [hunk 6]
```diff
@@ -51,9 +50,8 @@ The `message` property of the response body may contain additional details about
...
-| `400 "Bad request"`           | error details about rejected points, up to 100 points: `line` contains the first rejected line, `message` describes rejections line | If request data is malformed                                                                                                                                                                                                                                                                                                                                                     |
+| `400 "Bad request"`           | error details about rejected points, up to 100 points: `line` contains the first rejected line, `message` describes rejections      | If some (_When **partial writes** are configured for the cluster_) or all request data isn't allowed (for example, is malformed or falls outside of the bucket's retention period). The response body indicates if a partial write has occurred or all data has been rejected |
```
Comment (jstirnaman) [inline content/influxdb/cloud-dedicated/write-data/troubleshoot.md:54]:
```suggestion
| `400 "Bad request"`           | error details about rejected points, up to 100 points: `line` contains the first rejected line, `message` describes rejections      | If some (_when **partial writes** are configured for the cluster_) or all request data isn't allowed (for example, if it is malformed or falls outside of the bucket's retention period)--the response body indicates whether a partial write has occurred or if all data has been rejected |
```

File: content/influxdb/cloud-serverless/write-data/troubleshoot.md [hunk 7]
```diff
@@ -31,12 +31,12 @@ Learn how to avoid unexpected results and recover from errors when writing to {{
 
   1. Validates the request.
   2. If successful, attempts to [ingest data](/influxdb/cloud-serverless/reference/internals/durability/#data-ingest) from the request body; otherwise, responds with an [error status](#review-http-status-codes).
-  3. Ingests or rejects the entire batch and returns one of the following HTTP status codes:
+  3. Ingests or rejects data from the batch and returns one of the following HTTP status codes:
 
-     - `204 No Content`: some or all of the data is ingested and queryable
-     - `400 Bad Request`: all data is rejected
+     - `204 No Content`: all of the data is ingested and queryable
```
Comment (jstirnaman) [inline content/influxdb/cloud-serverless/write-data/troubleshoot.md:36]:
```suggestion
     - `204 No Content`: All of the data is ingested and queryable.
```

File: content/influxdb/cloud-serverless/write-data/troubleshoot.md [hunk 8]
```diff
@@ -31,12 +31,12 @@ Learn how to avoid unexpected results and recover from errors when writing to {{
...
+     - `400 Bad Request`: some or all of the data has been rejected. data which has not been rejected is ingested and queryable
```
Comment (jstirnaman) [inline content/influxdb/cloud-serverless/write-data/troubleshoot.md:37]:
```suggestion
     - `400 Bad Request`: Some or all of the data has been rejected. Data that has not been rejected is ingested and queryable.
```

File: content/influxdb/cloud-serverless/write-data/troubleshoot.md [hunk 9]
```diff
@@ -49,18 +49,16 @@ InfluxDB uses conventional HTTP status codes to indicate the success or failure
...
+| HTTP response code              | Response body                                                                                                                  | Description    |
+| :-------------------------------| :---------------------------------------------------------------                                                               | :------------- |
+| `204 "No Content"`              | no response body                                                                                                               | If InfluxDB ingested all of the data in the batch |
+| `400 "Bad request"`             | error details about rejected points, up to 100 points: `line` contains the first rejected line, `message` describes rejections | If some or all request data isn't allowed (for example, is malformed or falls outside of the bucket's retention period). The response body indicates if a partial write has occurred or all data has been rejected |
```
Comment (jstirnaman) [inline content/influxdb/cloud-serverless/write-data/troubleshoot.md:55]:
```suggestion
| `400 "Bad request"`             | error details about rejected points, up to 100 points: `line` contains the first rejected line, `message` describes rejections | If some or all request data isn't allowed (for example, is malformed or falls outside of the bucket's retention period)--the response body indicates whether a partial write has occurred or if all data has been rejected |
```

**Review by jstirnaman — CHANGES_REQUESTED** [review]
A few minor fixes. Thanks for updating this! Should we apply the Dedicated changes to Clustered as well?

File: api-docs/cloud-dedicated/v2/ref.yml (hunk 2 above)
Comment (savage-engineer) [reply api-docs/cloud-dedicated/v2/ref.yml:666]: Oops, good catch 🫣

File: content/influxdb/cloud-serverless/write-data/troubleshoot.md (hunk 9 above)
Comment (savage-engineer) [reply content/influxdb/cloud-serverless/write-data/troubleshoot.md:55]: Does `--` render as an em-dash or an en-dash in the docs site? I'm not particularly familiar with what renderer is in use!

**Comment by savage-engineer** [comment]
> A few minor fixes. Thanks for updating this! Should we apply the Dedicated changes to Clustered as well?

I don't think so yet. Reading https://github.com/influxdata/project-clustered/issues/593, it doesn't look like any clustered customers have switched on partial writes. The current docs don't seem to reference them either[^1]. 

I think it'll need documenting as part of the next clustered release _after_ we default to them on the 16th. Maybe one for @dgnorton to be aware of?

[^1]: https://docs.influxdata.com/influxdb/clustered/write-data/troubleshoot/#review-http-status-codes

**Comment by sanderson** [comment]
> I think it'll need documenting as part of the next clustered release after we default to them on the 16th. Maybe one for @dgnorton to be aware of?

Yes, that's the plan. After the next release, the partial writes will be the default behavior, but Clustered users will be able to disable partial writes using an environment variable.

Thanks for this work @savage-engineer. I started this on another branch last week but didn't make it very far, so we'll just use yours 😄.

I wasn't fully aware of the rollout timeline so I hadn't planned on staggering this change between dedicated and clustered, but makes sense that we do. I'll queue up these changes on a separate branch specifically for clustered.

**Comment by savage-engineer** [comment]
> > I think it'll need documenting as part of the next clustered release after we default to them on the 16th. Maybe one for @dgnorton to be aware of?
> 
> Yes, that's the plan. After the next release, the partial writes will be the default behavior, but Clustered users will be able to disable partial writes using an environment variable.
> 
> Thanks for this work @savage-engineer. I started this on another branch last week but didn't make it very far, so we'll just use yours 😄.
> 
> I wasn't fully aware of the rollout timeline so I hadn't planned on staggering this change between dedicated and clustered, but makes sense that we do. I'll queue up these changes on a separate branch specifically for clustered.

Sorry about the surprise on the rollout timeline - I should have reached out much sooner!

File: content/influxdb/cloud-serverless/write-data/troubleshoot.md (hunk 9 above)
Comment (jstirnaman) [reply content/influxdb/cloud-serverless/write-data/troubleshoot.md]: We render it as as em-dash b/c it's easier to type. It was a friendly point of contention in the past, but laziness won out. I'm not sure if that comes with the renderer or it's a custom style rule.

Comment (savage-engineer) [reply content/influxdb/cloud-serverless/write-data/troubleshoot.md]: Ah, very cool. I'll store that one in the toolbox 😎

**Comment by jstirnaman** [comment]
> > > I think it'll need documenting as part of the next clustered release after we default to them on the 16th. Maybe one for @dgnorton to be aware of?
> > 
> > 
> > Yes, that's the plan. After the next release, the partial writes will be the default behavior, but Clustered users will be able to disable partial writes using an environment variable.
> > Thanks for this work @savage-engineer. I started this on another branch last week but didn't make it very far, so we'll just use yours 😄.
> > I wasn't fully aware of the rollout timeline so I hadn't planned on staggering this change between dedicated and clustered, but makes sense that we do. I'll queue up these changes on a separate branch specifically for clustered.
> 
> Sorry about the surprise on the rollout timeline - I should have reached out much sooner!

We're grateful for the help!

**Review by jstirnaman — APPROVED** [review]

**Comment by dsprogis** [comment]
@jstirnaman , @sanderson , would you confirm that we are providing instructions for the user to be able to see the current configuration so that the user can confirm that they have successfully changed the partial writes variable?  @bnpfeife is putting a comment on Jacob's github issue with instructions.  Thanks!

**Pull request merged**